src/librustc_trans/trans/base.rs

```diff
@@ -848,7 +918,16 @@ pub fn from_arg_ty(bcx: Block, val: ValueRef, ty: Ty) -> ValueRef {
 
 pub fn to_arg_ty(bcx: Block, val: ValueRef, ty: Ty) -> ValueRef {
     if ty.is_bool() {
-        Trunc(bcx, val, Type::i1(bcx.ccx()))
+        let val = Value(val);
+        if let Some(zext) = val.as_zext_inst() {
+            if let Some(val) = zext.get_operand(0) {
+                let valty = val_ty(val.get());
+                if valty == Type::i1(bcx.ccx()) {
+                    return val.get();
```

Comment:
Is this worth the complexity? A extend/truncate pair seems like something that neither LLVM nor CPUs (for debug builds) should spend any significant amount of time on.

I guess I might have written the same thing, but looking at it, it just seems so complex.